File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/java/io/github/lishangbu/avalon/dataset/service/impl/BerryFirmnessServiceImplTest.java
package io.github.lishangbu.avalon.dataset.service.impl;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.BerryFirmness;
import io.github.lishangbu.avalon.dataset.repository.BerryFirmnessRepository;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class BerryFirmnessServiceImplTest {
    @Mock private BerryFirmnessRepository berryFirmnessRepository;
    @InjectMocks private BerryFirmnessServiceImpl berryFirmnessService;

    @Test
    void getPageByCondition_callsRepository() {
        BerryFirmness berryFirmness = new BerryFirmness();
        Pageable pageable = PageRequest.of(0, 5);
        Page<BerryFirmness> page = new PageImpl<>(List.of(berryFirmness));
        when(berryFirmnessRepository.findAll(any(Example.class), eq(pageable))).thenReturn(page);

        Page<BerryFirmness> result =
                berryFirmnessService.getPageByCondition(berryFirmness, pageable);

        assertSame(page, result);
        ArgumentCaptor<Example> captor = ArgumentCaptor.forClass(Example.class);
        verify(berryFirmnessRepository).findAll(captor.capture(), eq(pageable));
        assertNotNull(captor.getValue().getProbe());
    }

    @Test
    void listByCondition_callsRepository() {
        BerryFirmness berryFirmness = new BerryFirmness();
        List<BerryFirmness> expected = List.of(berryFirmness);
        when(berryFirmnessRepository.findAll(any(Example.class))).thenReturn(expected);

        List<BerryFirmness> result = berryFirmnessService.listByCondition(berryFirmness);

        assertSame(expected, result);
        verify(berryFirmnessRepository).findAll(any(Example.class));
    }

    @Test
    void save_usesRepository() {
        BerryFirmness berryFirmness = new BerryFirmness();
        when(berryFirmnessRepository.save(berryFirmness)).thenReturn(berryFirmness);

        BerryFirmness result = berryFirmnessService.save(berryFirmness);

        assertSame(berryFirmness, result);
        verify(berryFirmnessRepository).save(berryFirmness);
    }

    @Test
    void update_usesRepository() {
        BerryFirmness berryFirmness = new BerryFirmness();
        when(berryFirmnessRepository.save(berryFirmness)).thenReturn(berryFirmness);

        BerryFirmness result = berryFirmnessService.update(berryFirmness);

        assertSame(berryFirmness, result);
        verify(berryFirmnessRepository).save(berryFirmness);
    }

    @Test
    void removeById_callsRepository() {
        berryFirmnessService.removeById(1L);
        verify(berryFirmnessRepository).deleteById(1L);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/kotlin/io/github/lishangbu/avalon/dataset/service/impl/BerryFirmnessServiceImplTest.kt
package io.github.lishangbu.avalon.dataset.service.impl

import io.github.lishangbu.avalon.dataset.entity.BerryFirmness
import io.github.lishangbu.avalon.dataset.entity.dto.BerryFirmnessSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.BerryFirmnessView
import io.github.lishangbu.avalon.dataset.entity.dto.SaveBerryFirmnessInput
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateBerryFirmnessInput
import io.github.lishangbu.avalon.dataset.repository.BerryFirmnessRepository
import org.babyfish.jimmer.Page
import org.babyfish.jimmer.sql.ast.mutation.AssociatedSaveMode
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import org.springframework.data.domain.PageRequest

class BerryFirmnessServiceImplTest {
    private val repository = mock(BerryFirmnessRepository::class.java)
    private val service = BerryFirmnessServiceImpl(repository)

    @Test
    fun getPageByCondition_callsRepository() {
        val specification = BerryFirmnessSpecification(id = "1", internalName = "hard")
        val pageable = PageRequest.of(0, 5)
        `when`(repository.pageViews(specification, pageable)).thenReturn(Page(listOf(berryFirmnessView(1L)), 1, 1))

        val result = service.getPageByCondition(specification, pageable)

        assertEquals(1, result.rows.size)
        assertEquals("1", result.rows.first().id)
    }

    @Test
    fun listByCondition_callsRepository() {
        val specification = BerryFirmnessSpecification(id = "1", internalName = "hard")
        `when`(repository.listViews(specification)).thenReturn(listOf(berryFirmnessView(1L)))

        val result = service.listByCondition(specification)

        assertEquals(1, result.size)
        assertEquals("1", result.first().id)
    }

    @Test
    fun save_usesInsertOnlyMode() {
        `when`(repository.save(any<BerryFirmness>(), eq(SaveMode.INSERT_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())).thenReturn(berryFirmnessEntity(1L))

        val result = service.save(SaveBerryFirmnessInput("hard", "硬"))

        assertEquals("1", result.id)
        verify(repository).save(any<BerryFirmness>(), eq(SaveMode.INSERT_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())
    }

    @Test
    fun update_usesUpsertMode() {
        `when`(repository.save(any<BerryFirmness>(), eq(SaveMode.UPDATE_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())).thenReturn(berryFirmnessEntity(1L))

        val result = service.update(UpdateBerryFirmnessInput("1", "hard", "硬"))

        assertEquals("1", result.id)
        verify(repository).save(any<BerryFirmness>(), eq(SaveMode.UPDATE_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())
    }

    @Test
    fun removeById_callsRepository() {
        service.removeById(1L)

        verify(repository).deleteById(1L)
    }
}

private fun berryFirmnessEntity(id: Long): BerryFirmness =
    BerryFirmness {
        this.id = id
        internalName = "hard"
        name = "硬"
    }

private fun berryFirmnessView(id: Long): BerryFirmnessView = BerryFirmnessView(berryFirmnessEntity(id))
